adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -1274,6 +1321,119 @@ protected final void onSubscribe(final MqttSubscribeMessage subscribeMsg) {
+
+            final int errorCode = ServiceInvocationException.extractStatusCode(error);
+            final String clientFacingErrorMessage = error instanceof ServerErrorException
+                    ? ((ServerErrorException) error).getClientFacingMessage()
+                    : null;
+            final String errorMessage = Optional.ofNullable(clientFacingErrorMessage).orElse(error.getMessage());
+            final String correlationId = Optional.ofNullable(context.correlationId()).orElse("-1");
+            final String publishTopic = subscription.getErrorPublishTopic(context, errorCode);
+            Tags.MESSAGE_BUS_DESTINATION.set(span, publishTopic);
+            TracingHelper.TAG_QOS.set(span, subscription.getQos().name());
+
+            final JsonObject errorJson = new JsonObject();
+            errorJson.put("code", errorCode);
+            errorJson.put("message", errorMessage);
+            errorJson.put("timestamp", ZonedDateTime.now(ZoneOffset.UTC).truncatedTo(ChronoUnit.MILLIS)
+                    .format(DateTimeFormatter.ISO_INSTANT));
+            errorJson.put(MessageHelper.SYS_PROPERTY_CORRELATION_ID, correlationId);
+
+            final String targetInfo = authenticatedDevice != null && !authenticatedDevice.getDeviceId().equals(context.deviceId())
+                    ? String.format("gateway [%s], device [%s]", authenticatedDevice.getDeviceId(), context.deviceId())
+                    : String.format("device [%s]", context.deviceId());
+
+            final Promise<Integer> resultPromise = Promise.promise();
+            endpoint.publish(publishTopic, errorJson.toBuffer(), subscription.getQos(), false, false, resultPromise);
+            return resultPromise.future()
+                    .onSuccess(msgId -> {
+                        log.debug("published error message [packet-id: {}] to {} [tenant-id: {}, MQTT client-id: {}, QoS: {}, topic: {}]",
+                                msgId, targetInfo, subscription.getTenant(), endpoint.clientIdentifier(),
+                                subscription.getQos(), publishTopic);
+                        span.log(subscription.getQos().value() > 0 ? "published error message, packet-id: " + msgId
+                                : "published error message");
+                    }).onFailure(thr -> {
+                        log.debug("error publishing error message to {} [tenant-id: {}, MQTT client-id: {}, QoS: {}, topic: {}]",
+                                targetInfo, subscription.getTenant(), endpoint.clientIdentifier(),
+                                subscription.getQos(), publishTopic, thr);
+                        TracingHelper.logError(span, "failed to publish error message", thr);
+                    }).compose(msgId -> {
+                        if (MqttQoS.AT_LEAST_ONCE.equals(subscription.getQos())) {
+                            final Promise<Void> ackOrTimeoutPromise = Promise.promise();
+                            pendingAcks.add(
```

Comment:
We are adding to `pendingAcks` once the message has been successfully sent. I think there is still chance that a `PUBACK` has already arrived even before adding to `pendingAcks`. In this case, we will wait for `PUBACK` and then time out. Even though we have a timeout mechanism, but this is misleading as it seems no `PUBACK` has arrived. The same behavior is also there while we publish commands.

I think it would be appropriate to add to the `pendingAcks` first and then publish the message using the below method. In this case, we have to generate the message identifier, add it to `pendingAcks` followed by publishing the message. We can also handle this in a separate PR.

```
MqttEndpoint publish(String topic, Buffer payload, MqttQoS qosLevel, boolean isDup, boolean isRetain, int messageId, Handler<AsyncResult<Integer>> publishSentHandler);
```

Reply:
I have tested the above scenario with introducing some extra statements so that after successfully publishing an error message it takes some time till `pendingAcks.add(` is called. I have noticed that the `PUBACKs` have been always processed by the same event-loop thread. So we wouldn't end up with the above situation of missing `PUBACKs`

Reply:
The above "publish()" invocation will complete the result promise directly (without the event loop thread calling other handlers in between). Therefore it is safe to invoke "pendingAcks.add" here since there can't be an incoming PUBACK getting handled between "publish" and "pendingAcks.add". This is also because only one event loop thread is involved in the handling of requests (like SUBSCRIBE/PUBACK) for the MQTT connection here (`AbstractVertxBasedMqttProtocolAdapter` is a verticle).

This is of course somewhat hidden in the `MqttEndpointImpl.publish` implementation. Therefore things could be made more obvious by doing the `pendingAcks.add` invocation *before* the *publish* invocation. I would leave that as something that can be done in a subsequent PR.